### PR TITLE
Enable fused softmax smoothing

### PR DESCRIPTION
**Pull request by KellenSunderland**
MXNet modifications to support softmax smoothing. 

@larroy @fhieber @tdomhan

### CODE REVIEW

**Comment by tdomhan** [comment]
thanks! make sure to close this issue after submiting: https://github.com/apache/incubator-mxnet/issues/6996

File: src/operator/softmax_output-inl.h [hunk 1]
```diff
@@ -78,6 +79,10 @@ struct SoftmaxOutputParam : public dmlc::Parameter<SoftmaxOutputParam> {
     DMLC_DECLARE_FIELD(out_grad)
     .set_default(false)
     .describe("Multiplies gradient with output gradient element-wise.");
+    DMLC_DECLARE_FIELD(smooth_alpha)
+    .set_default(0.0f)
+    .set_range(0.0f,1.0f)
+    .describe("Constant for smoothed cross-entropy gradients.");
```
Comment (tdomhan) [inline src/operator/softmax_output-inl.h]: A little more detail for this parameter could make sense:

> "Constant for computing a label smoothed version of cross-entropy for the backwards pass, in which this constant gets subtracted from the one-hot encoding of the gold label and distributed uniformly to all other labels."
Something along these lines.

File: tests/python/unittest/test_operator.py [hunk 2]
```diff
@@ -235,6 +235,55 @@ def test_regression():
...
+    grad_x = mx.nd.zeros((1,4), ctx=default_context())
+    label_nd = mx.nd.array([1], ctx=default_context())
+
+    sym = mx.sym.SoftmaxOutput(data=x, label=label, ignore_label=0, use_ignore=False)
+    ex = sym.bind(ctx=default_context(), args={'x': x_nd, 'label': label_nd}, args_grad={'x': grad_x})
+
+    ex.forward(is_train=True)
+    softmax_out = ex.outputs[0].asnumpy()
+    expected_softmax_out = mx.nd.softmax(x_nd).asnumpy()
+    print(softmax_out)
```
Comment (tdomhan) [inline tests/python/unittest/test_operator.py]: print statement

File: tests/python/unittest/test_operator.py [hunk 3]
```diff
@@ -235,6 +235,55 @@ def test_regression():
...
+    x_nd = mx.nd.array([[1, 6, 4, 2]], ctx=default_context())
+    grad_x = mx.nd.zeros((1,4), ctx=default_context())
+    label_nd = mx.nd.array([1], ctx=default_context())
+
+    sym = mx.sym.SoftmaxOutput(data=x, label=label, ignore_label=0, use_ignore=False)
+    ex = sym.bind(ctx=default_context(), args={'x': x_nd, 'label': label_nd}, args_grad={'x': grad_x})
+
+    ex.forward(is_train=True)
+    softmax_out = ex.outputs[0].asnumpy()
+    expected_softmax_out = mx.nd.softmax(x_nd).asnumpy()
```
Comment (tdomhan) [inline tests/python/unittest/test_operator.py]: it'd probably make sense to computed the expected softmax by hand. Who is to say that mx.nd.softmax and mx.sym.SoftMax don't share any code under the hood. (I guess they don't but they might in the future)

**Comment by KellenSunderland** [comment]
Will do.

File: tests/python/unittest/test_operator.py [hunk 4]
```diff
@@ -235,6 +235,55 @@ def test_regression():
...
+    grad_x = mx.nd.zeros((1,4), ctx=default_context())
+    label_nd = mx.nd.array([1], ctx=default_context())
+
+    sym = mx.sym.SoftmaxOutput(data=x, label=label, ignore_label=0, use_ignore=False, smooth_alpha=alpha)
+    ex = sym.bind(ctx=default_context(), args={'x': x_nd, 'label': label_nd}, args_grad={'x': grad_x})
+
+    ex.forward(is_train=True)
+    softmax_out = ex.outputs[0].asnumpy()
+    expected_softmax_out = mx.nd.softmax(x_nd).asnumpy()
+    print(softmax_out)
```
Comment (tdomhan) [inline tests/python/unittest/test_operator.py]: another print

File: tests/python/unittest/test_operator.py [hunk 5]
```diff
@@ -235,6 +235,55 @@ def test_regression():
                      lambda x, y : x - y)
 
 
+def check_softmax_grad():
```
Comment (tdomhan) [inline tests/python/unittest/test_operator.py:238]: where is this called from?

File: tests/python/unittest/test_operator.py [hunk 6]
```diff
@@ -235,6 +235,55 @@ def test_regression():
...
+    ex.forward(is_train=True)
+    softmax_out = ex.outputs[0].asnumpy()
+    expected_softmax_out = mx.nd.softmax(x_nd).asnumpy()
+    print(softmax_out)
+    assert np.isclose(softmax_out, expected_softmax_out).all()
+
+    ex.backward(is_train=True)
+    grad_out = ex.grad_arrays[0].asnumpy()
+    k = int(label_nd[0].asscalar())
+    expected_grad_out = np.full((1,4), fill_value=-alpha/(4-1))
```
Comment (tdomhan) [inline tests/python/unittest/test_operator.py]: probably makes sense to wrap the `4-1` in `float(4-1)`.

File: tests/python/unittest/test_operator.py (hunk 3 above)
Comment (KellenSunderland) [reply tests/python/unittest/test_operator.py]: Good idea, will do.

File: src/operator/softmax_output-inl.h (hunk 1 above)
Comment (KellenSunderland) [reply src/operator/softmax_output-inl.h]: Great improvement, updated.

File: tests/python/unittest/test_operator.py (hunk 2 above)
Comment (KellenSunderland) [reply tests/python/unittest/test_operator.py]: fixed.

File: tests/python/unittest/test_operator.py (hunk 4 above)
Comment (KellenSunderland) [reply tests/python/unittest/test_operator.py]: fixed.

File: tests/python/unittest/test_operator.py (hunk 6 above)
Comment (KellenSunderland) [reply tests/python/unittest/test_operator.py]: done.
